Blog on mpOpenApi (#4296)
OpenLiberty/blogs: posts/2025-03-28-Building-Better-APIs:-Documenting-APIs-in-Open-Liberty-using-mpOpenAPI.adoc

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
additional_authors: 
- name: Abhinav K
  github: https://github.com/itsabhinavk
  image: https://avatars0.githubusercontent.com/itsabhinavk
---
= Building Better APIs: Documenting APIs in Open Liberty using mpOpenAPI
Sweetty P Devassy <https://github.com/sweettypdevassy>
:imagesdir: /
:url-prefix:
:url-about: /

In 2012, Twitter made significant changes to its API policies and access rules, tightening control over how third-party developers could interact with its platform. These changes affected many apps that relied on Twitter's API for services like tweet posting, timeline retrieval, and user authentication. However, due to a lack of proper API documentation, several third-party apps using Twitter’s API were unaware of these changes, leading to overnight crashes. This incident, along with similar issues across other platforms, acted as a wake-up call for the industry to standardise API documentation. 

> With evolving technology, we are building a huge number of highly complicated APIs. Documenting these APIs requires effort and is often seen as a tedious task. Furthermore, complex APIs make documentation even more challenging. In reality, API documentation serves as a means of sharing up-to-date information between development teams. This crucial aspect later led to the development of Swagger and later OpenAPI specification. So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with the mpOpenAPI library. 

Reviewer: This is a nice story, but how sure are we of its accuracy?

1. Would proper API documentation have avoided the issues? You say that the changes were around "API policies and access rules", not in the API definition itself. It sounds more like a communication and policy issue, rather than one that would be fixed by the type of API documentation provided by OpenAPI.
2. Did this definitely act as a wake up call to standardise API documentation? Do you have evidence or personal experience of this?
3. These paragraphs suggest that the twitter API changes lead eventually to the development of Swagger and OpenAPI, yet the Wikipedia page says the Swagger project open sourced its code in 2011, which is backed up by the history of the swagger-core repository.

Reviewer 2: The reason why we took up this story was because this incident shows importance of API documentation. Although Twitter did announce that they were upgrading to API v1.1, some people got confused about the changes.
-   [Ref](https://www.theverge.com/2012/8/23/3263481/twitter-api-third-party-developers) to changes Twitter made in 2012 (scroll down a bit to reach articles from 2012).
- A [stack overflow question](https://stackoverflow.com/questions/17143985/twitter-api-error-215) from a user that shows some issues Twitter API consumers faced while migrating to v1.1.

These references show that its not easy to understand changes in API by reading a long changelist article, and maybe a better way of documenting and publishing API documentation could be a better approach.

This story is just an example at the beginning of our article to show importance of proper documentation and to engage readers interest.

Reviewer: Well, it's your blog post so I'll leave it up to you, but I am a little concerned that you're exaggerating a bit here and elsewhere in the post.

1. Good API documentation alone would not have fixed the twitter problem. Consequently I wouldn't say that the twitter problem was a wake-up call to the industry that we needed better API documentation.
    - frankly, the twitter thing was a wake-up call that you can't trust that any company is going to maintain an API that you rely on, especially if you don't have a contract with them and aren't paying for it
2. I wouldn't use the word "effortless" after describing steps that the user has to do. You could perhaps say that the automatic generation without annotations is "effortless", but writing good documentation always involves effort, we just want to minimise it.
3. I think "it would be hard to imagine _What would software development look like today if OpenAPI never existed?_" is a bit of a stretch. I think OpenAPI has had a big impact on the use and documentation of RESTful APIs, but I don't think you can say software development as a whole would look drastically different without it.

I'll not comment further on this, it's your blog post with your name on it so I'll just focus on the accuracy of the content which relates to Open Liberty and MicroProfile OpenAPI.

Author: Revised as per the suggestions.
e2427639914e2e2e9ea6aa562dc01674f6612238
1063082669fbb4188f8ea92bd6fb28dfad4b2573
0b68ce38b48b9cba0db44adc060601c2c91265bf

> With evolving technology, we are building a huge number of highly complicated APIs. Documenting these APIs requires effort and is often seen as a tedious task. Furthermore, complex APIs make documentation even more challenging. In reality, API documentation serves as a means of sharing up-to-date information between development teams. This crucial aspect later led to the development of Swagger and later OpenAPI specification. So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with the mpOpenAPI library. 

Reviewer: > mpOpenAPI library

mpOpenAPI library is not the correct terminology here.

Liberty has MicroProfile OpenAPI features (one for each version of the MicroProfile OpenAPI spec) These features have short names which start with mpOpenAPI.

Author: Changed to MicroProfile OpenAPI features in https://github.com/OpenLiberty/blogs/pull/4296/commits/7d4f374f0a0237ed1636f5e77013c053754c16bc.

Reviewer: There are still a lot of places where you're saying "mpOpenAPI" without saying what that means.

There's three things with confusingly similar names:
- OpenAPI - the specification for describing RESTful APIs in a JSON or YAML format (you describe this one well)
- MicroProfile OpenAPI - the specification describing for creating and serving OpenAPI documents based on a Jakarta REST application
- Open Liberty's MicroProfile OpenAPI features - which implement the MicroProfile OpenAPI specification

In most cases, users may not care about the difference between the last two, but when we're writing about it, we should try to be consistent and correct.

When we're talking about the annotations, those come from MicroProfile OpenAPI. When we're talking about the feature that the user enables or what Open Liberty provides, we're usually talking about the feature.

In general I think it's fine to abbreviate MicroProfile to MP after using it once.

Most of the issues are in line 49, I'll add some suggestions to that part of the document. I'll try to catch any other places you've used `mpOpenAPI` but do check if I've missed any or if I've written something inconsistent.

Author: Changed to MicroProfile OpenAPI
a4b8c4c32e4791b2aaa5e9b7358022a7db07ae2c
7ccff2f7c673165493e212436deffba42b3c2cc3
8e826012c3ae6c9faef46812c452a5122b3d423d

== Adding mpOpenAPI to Open Liberty 

To add OpenAPI to Open Liberty, we need to use MicroProfile and enable the mpOpenAPI feature. 

1 . Add MicroProfile dependency in `pom.xml`
[source, xml]
----
<dependency> 
  <groupId>org.eclipse.microprofile</groupId> 
  <artifactId>microprofile</artifactId> 
  <version>7.0</version> 
  <type>pom</type> 
  <scope>provided</scope> 
</dependency> 
----

2 . Update `server.xml` in `liberty/config` to enable mpOpenAPI: 
[source, xml]
----
<featureManager> 
    <feature>jakartaee-10.0</feature> 
    <feature>mpOpenAPI-4.0</feature> 
</featureManager>

> ----

Reviewer: You're using the POM dependency for MicroProfile 7.0, but the `mpOpenAPI-4.0` feature.

I would suggest you either use the `microprofile-7.0` feature, or use the `microprofile-open-api` POM dependency.

Author: changed to microprofile-7.0 feature in https://github.com/OpenLiberty/blogs/pull/4296/commits/7d4f374f0a0237ed1636f5e77013c053754c16bc.

Reviewer: This looks good, may want to add a note that `microprofile-7.0` includes the `mpOpenAPI-4.0` feature.

Author: Updated
efdf2e05786d4c70f2c8dead3c4e9860d740a44a

3 . Create an Endpoints.java file, 

* This is where you will define your API endpoints using OpenAPI annotations. 

> * The full source code for this example is available on GitHub: link:https://github.com/sweettypdevassy/openAPI-blog[GitHub Repository.] 

Reviewer: It's not clear whether these are steps for the user to follow along with, using the example project provided, or generic steps for using MP OpenAPI in an existing project.

If the former it might be better to start with the example repository before listing the steps the user is to follow.

If the latter then the steps should be less specific.

It would be fine to list general steps, and then show an example, and then say you'll use the example project for the rest of the blog.

Author: added general steps, followed by an example, and stated that this example project will be used throughout the rest of the blog. 
Commit: https://github.com/OpenLiberty/blogs/pull/4296/commits/7d4f374f0a0237ed1636f5e77013c053754c16bc

Reviewer: The steps are clearer now, thanks :+1:

4 .  Run the project using Maven: `mvn liberty:dev`

5 . Test using Swagger UI: 

* Open: http://localhost:9081/openapi/ui/ 
 
This is how it appears on the Swagger UI

[.img_border_light]
image::img/blog/OPENAPI-1.png[Swagger UI ,width=70%,align="center"]

== OpenAPI Annotations 

mpOpenAPI provides the following annotations:

[.img_border_light]

> image::img/blog/OPENAPI-2.png[OpenAPI Annotations ,width=70%,align="center"] 

Reviewer: Textual information should be included as text, not in an image.

Author: changed the image into a table in https://github.com/OpenLiberty/blogs/pull/4296/commits/7d4f374f0a0237ed1636f5e77013c053754c16bc.

Below is an example of using MicroProfile OpenAPI annotations to define an API, its metadata, security, and operations. 

*1 . API Metadata Definitions*

> * *`@OpenAPIDefinition`* - Defines metadata for the entire API.

Reviewer: It's probably helpful to point out that basic documentation is generated from REST methods automatically, and annotations are generally used to enhance it.

You can draw a parallel with Javadoc comments. You can generate Javadoc for a class with no javadoc comments, it will list the methods and the parameter names and return types, but won't say how they behave.

OpenAPI annotations are mostly used in the same way as Javadoc, to provide explanation of the REST methods and how to use them.

The other reason they're used is to provide information that can't be picked up automatically from the Java code for the REST method.

Author: added a  section in https://github.com/OpenLiberty/blogs/pull/4296/commits/8652653eff0eadb61deb33e5e46d0d44c0b2023f

* *`@Info`* - Provides metadata such as title, description, version, and contact details.  

* *`@Server`* - Specifies one or more servers where the API is hosted.  

* *`@Tag`* - Defines logical groupings of endpoints.  

> * *`@SecurityRequirement`* - Enforces security (e.g., JWT authentication) for API endpoints.  

Reviewer: `@SecurityRequirement` does not enforce security. It documents that security is enforced.

If you add this without correctly setting up security, then your documentation will say that your API is secured, but it won't be.

Author: edited the article to describe this annotation as a documentation of security in https://github.com/OpenLiberty/blogs/pull/4296/commits/7d4f374f0a0237ed1636f5e77013c053754c16bc.

[source,  java]
----
@OpenAPIDefinition( 
    info = @Info( 
        title = "Blog API", 
        version = "1.0", 
        description = "This is an example API using MicroProfile OpenAPI", 
        contact = @Contact(name = "Sweetty", email = "sweetty@gmail.com"), 
        license = @License(name = "Apache 2.0", url = "https://apache.org/licenses/LICENSE-2.0.html") 
    ), 
    servers = @Server(url = "http://localhost:9081/openapi-demo", description = "Localhost Server"), 
    tags = { 
        @Tag(name = "Blog", description = "Operations related to blog management") 
    }, 
    security = @SecurityRequirement(name = "jwtAuth") 
) 

> ----

Reviewer: Any example with annotations needs to show where the annotation is being placed.

Author: ![image](https://github.com/user-attachments/assets/35625b21-0c47-4407-9235-126d07f2bb6a)
Added highlights, It is highlighted in the VSCode preview but not in the GitHub preview.

Reviewer: Sorry, this isn't what I meant.

When an annotation is used, it's attached to the declaration of some element in the source code, for example a class, a method, a field or a parameter.

I would suggest that your examples should be clearer about _where_ the user should put the annotation. `@OpenAPIDefinition` should be used on a class (usually the REST `Application` class, but I don't think that's a requirement), whereas the annotations which apply to an operation should be put on a REST resource method (which will also be annotated with REST annotations, e.g. `@GET`, or `@POST`).

I would add some text explaining where the annotations should be used, and include the element they're attached to in the example.

E.g.
```java
@OpenAPIDefinition( 
    info = @Info( 
        title = "Blog API", 
        version = "1.0", 
        description = "This is an example API using MicroProfile OpenAPI", 
        contact = @Contact(name = "Sweetty", email = "sweetty@gmail.com"), 
        license = @License(name = "Apache 2.0", url = "https://apache.org/licenses/LICENSE-2.0.html") 
    ), 
    servers = @Server(url = "http://localhost:9081/openapi-demo", description = "Localhost Server"), 
    tags = { 
        @Tag(name = "Blog", description = "Operations related to blog management") 
    }, 
    security = @SecurityRequirement(name = "jwtAuth") 
)
public class MyApplication extends Application {
```

Author: updated
917c103609676a0963f06ec522f5e16d3daf3bb7

* Inspired by OpenAPI's *Swagger UI*, gRPC now has *gRPC UI*, which allows developers to explore and test API methods interactively, much like Swagger does for REST APIs.   

== Conclusion  
In today's world of *microservices and cloud computing,* standardised API documentation is more important than ever. OpenAPI has become the foundation for creating *consistent, accessible, and well-documented APIs,* ensuring seamless integration across diverse systems. As APIs continue to evolve, having a structured approach to documentation helps developers *improve collaboration, reduce errors, and accelerate development cycles.* For those working with *Open Liberty,* exploring *mpOpenAPI* offers an effortless way to generate OpenAPI-compliant documentation, making API management more efficient. As we embrace automation and interoperability in modern software development, it would be hard to imagine _What would software development look like today if OpenAPI never existed?_

OpenLiberty.io site links:

link:https://openliberty.io/guides/microprofile-openapi.html[Documenting RESTful APIs]

> link:https://openliberty.io/docs/23.0.0.3/documentation-openapi.html[API documentation with OpenAPI]

Reviewer: Should we link to a more up to date version of liberty here?

Author: added more up to date version of liberty.

---
layout: post

> title: "Building Better APIs: Documenting APIs in Open Liberty using mpOpenAPI"
> categories: blog
> author_picture: https://avatars3.githubusercontent.com/sweettypdevassy
> author_github: https://github.com/sweettypdevassy
> seo-title: Building Better APIs, Documenting APIs in Open Liberty using mpOpenAPI - OpenLiberty.io
> seo-description: This blog explores the OpenAPI Specification and its integration with Open Liberty using mpOpenAPI. It covers the importance of API documentation, the role of OpenAPI and Swagger, and a step-by-step guide to adding mpOpenAPI to an Open Liberty project. Readers will learn about various mpOpenAPI annotations with practical examples and see how they are rendered in Swagger UI.
> blog_description: "This blog explores the OpenAPI Specification and its integration with Open Liberty using mpOpenAPI. It covers the importance of API documentation, the role of OpenAPI and Swagger, and a step-by-step guide to adding mpOpenAPI to an Open Liberty project. Readers will learn about various mpOpenAPI annotations with practical examples and see how they are rendered in Swagger UI."
> open-graph-image: https://openliberty.io/img/twitter_card.jpg
> open-graph-image-alt: Open Liberty Logo
> additional_authors: 
> - name: Abhinav K
>   github: https://github.com/itsabhinavk
>   image: https://avatars0.githubusercontent.com/itsabhinavk
> ---
> = Building Better APIs: Documenting APIs in Open Liberty using mpOpenAPI

Reviewer: ```suggestion
title: "Building Better APIs: Documenting APIs in Open Liberty using MicroProfile OpenAPI"
categories: blog
author_picture: https://avatars3.githubusercontent.com/sweettypdevassy
author_github: https://github.com/sweettypdevassy
seo-title: Building Better APIs, Documenting APIs in Open Liberty using MicroProfile OpenAPI - OpenLiberty.io
seo-description: This blog explores the OpenAPI Specification and its integration with Open Liberty using MicroProfile OpenAPI. It covers the importance of API documentation, the role of OpenAPI and Swagger, and a step-by-step guide to using MicroProfile OpenAPI in an Open Liberty project. Readers will learn about various MicroProfile OpenAPI annotations with practical examples and see how they are rendered in Swagger UI.
blog_description: "This blog explores the OpenAPI Specification and its integration with Open Liberty using MicroProfile OpenAPI. It covers the importance of API documentation, the role of OpenAPI and Swagger, and a step-by-step guide to using MicroProfile OpenAPI in an Open Liberty project. Readers will learn about various MicroProfile OpenAPI annotations with practical examples and see how they are rendered in Swagger UI."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
additional_authors: 
- name: Abhinav K
  github: https://github.com/itsabhinavk
  image: https://avatars0.githubusercontent.com/itsabhinavk
---
= Building Better APIs: Documenting APIs in Open Liberty using MicroProfile OpenAPI
```

</featureManager>
----
. Create an Endpoints.java file, 
* Define your API endpoints using OpenAPI annotations. For example:
+
[source,java]
----
import org.eclipse.microprofile.openapi.annotations.Operation;
import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("/hello")
public class Endpoints {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Operation(summary = "Returns a greeting message")
    @APIResponse(responseCode = "200", description = "Successful response")
    public String hello() {
        return "Hello, OpenAPI with Liberty!";
    }
}
----

. Run the project using Maven
* Execute: `mvn liberty:dev`

. Test using Swagger UI
* Open: http://localhost:9081/openapi/ui/

For a fully functional example, check out this link:https://github.com/sweettypdevassy/openAPI-blog[GitHub Repository]. This blog will use this example project for further explanations. 

This is how it appears on the Swagger UI
[.img_border_light]
image::img/blog/OPENAPI-1.png[Swagger UI ,width=70%,align="center"]

== OpenAPI Annotations 

> mpOpenAPI provides the following annotations:

Reviewer: ```suggestion
MicroProfile OpenAPI provides the following annotations:
```

While basic API documentation is automatically generated from REST methods in Open Liberty, OpenAPI annotations help enhance and refine it. This is similar to how Javadoc comments work in Java—without them, Javadoc can still list methods, parameters, and return types, but it lacks detailed explanations of behavior and usage. Likewise, OpenAPI annotations provide essential context to REST endpoints, describing their purpose, expected inputs, responses, and error handling. Additionally, some API details cannot be inferred directly from Java code, such as custom response headers, security requirements, or detailed descriptions of parameters. In such cases, annotations become essential for ensuring comprehensive, well-structured API documentation that is both developer-friendly and machine-readable.

These annotations play a crucial role in API development. They provide a clear, structured, and machine-readable way to define API behavior without requiring additional configuration files. These annotations allow developers to embed metadata directly into the code, ensuring that API endpoints, request parameters, response types, and security schemes are documented and consistent. 

> Annotations make mpOpenAPI integration easier and allow for automatic creation of documentation that complies with OpenAPI. This not only improves developer productivity but also enhances API discoverability, making it easier for consumers to understand and interact with the API. With help of these annotations, teams can ensure that their APIs remain maintainable, scalable, and aligned with industry standards, reducing manual documentation efforts and minimizing discrepancies between implementation and documentation. 

Reviewer: > Annotations make mpOpenAPI integration easier

Not sure this is quite what the annotations do.

Reviewer: Ok, this paragraph has some good ideas, but is a bit muddled. E.g. I'm unsure how annotation enhance API discoverability and make it easier for consumers to understand and interact with the API.

Here's what I think should be captured here:
- The MP OpenAPI feature generates the documentation.
- The MP OpenAPI annotations allow that documentation to be enhanced.
  - Having comprehensive documentation is helpful to end users
- The annotations are stored in the code, alongside the implementation.
  - This makes it easy for developers to add and maintain the documentation as they update the code

Author: Updated
17119b5cf6e840b3e8c406b34ca45d77665da620

== Influence  
As *API architectures evolve,* OpenAPI, the industry-standard specification for *documenting REST APIs,* is influencing how *GraphQL* and *gRPC* APIs handle documentation and interoperability. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability. 

* OpenAPI’s structured *JSON/YAML format* has influenced *GraphQL’s SDL (Schema Definition Language),* where types, queries, and mutations are explicitly defined. 

* Inspired by OpenAPI's *Swagger UI*, gRPC now has *gRPC UI*, which allows developers to explore and test API methods interactively, much like Swagger does for REST APIs.   

== Conclusion  

> In today's world of *microservices and cloud computing,* standardised API documentation is more important than ever. OpenAPI has become the foundation for creating *consistent, accessible, and well-documented APIs,* ensuring seamless integration across diverse systems. As APIs continue to evolve, having a structured approach to documentation helps developers *improve collaboration, reduce errors, and accelerate development cycles.* For those working with *Open Liberty,* exploring *mpOpenAPI* offers an effortless way to generate OpenAPI-compliant documentation, making API management more efficient. As we embrace automation and interoperability in modern software development, it would be hard to imagine _What would software development look like today if OpenAPI never existed?_

Reviewer: ```suggestion
In today's world of *microservices and cloud computing,* standardised API documentation is more important than ever. OpenAPI has become the foundation for creating *consistent, accessible, and well-documented APIs,* ensuring seamless integration across diverse systems. As APIs continue to evolve, having a structured approach to documentation helps developers *improve collaboration, reduce errors, and accelerate development cycles.* For those working with *Open Liberty,* exploring *MicroProfile OpenAPI* offers an effortless way to generate OpenAPI-compliant documentation, making API management more efficient. As we embrace automation and interoperability in modern software development, it would be hard to imagine _What would software development look like today if OpenAPI never existed?_
```

Author: Updated
7ccff2f7c673165493e212436deffba42b3c2cc3

== Adding MicroProfile OpenAPI to Open Liberty 

. Set Up Java
  * Install JDK 8 or later.
  * Set the `JAVA_HOME` environment variable.
    ** Linux/macOS: `export JAVA_HOME=/path/to/jdk`
    ** Windows: `set JAVA_HOME=C:\path\to\jdk`

. Download a sample Open Liberty Project
  * Go to the link:https://openliberty.io/start/[Open Liberty starter page].
  * Choose the appropriate release and download the `.zip` file.

. Extract the files
  * Extract the downloaded `.zip` file to your preferred directory.

> . Add MicroProfile dependency in `pom.xml`(microprofile-7.0 includes mpOpenAPI-4.0 feature)
> +
> [source,xml]
> ----
> <dependency>
>     <groupId>org.eclipse.microprofile</groupId>
>     <artifactId>microprofile</artifactId>
>     <version>7.0</version>
>     <type>pom</type>
>     <scope>provided</scope>
> </dependency>
> ----
> . Update `server.xml` in `liberty/config` to enable MicroProfile:
> +
> [source,xml]
> ----
> <featureManager>
>     <feature>jakartaee-10.0</feature>
>     <feature>microprofile-7.0</feature>
> </featureManager>
> ----

Reviewer: ```suggestion
. Add MicroProfile dependency in `pom.xml`
+
[source,xml]
----
<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>7.0</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>
----
. Update `server.xml` in `liberty/config` to enable MicroProfile: (`microprofile-7.0` includes `mpOpenAPI-4.0` feature)
+
[source,xml]
----
<featureManager>
    <feature>jakartaee-10.0</feature>
    <feature>microprofile-7.0</feature>
</featureManager>
----
```

. Create an Endpoints.java file, 

> * Define your API endpoints using OpenAPI annotations. For example:

Reviewer: ```suggestion
* Define your API endpoints and add MicroProfile OpenAPI annotations. For example:
```

Technically the Jakarta REST annotations define the endpoint. The MP OpenAPI annotations just document it.

+
[source,java]
----
import org.eclipse.microprofile.openapi.annotations.Operation;
import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;

@Path("/hello")
public class Endpoints {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Operation(summary = "Returns a greeting message")
    @APIResponse(responseCode = "200", description = "Successful response")
    public String hello() {
        return "Hello, OpenAPI with Liberty!";
    }
}
----

. Run the project using Maven
* Execute: `mvn liberty:dev`

. Test using Swagger UI

> * Open: http://localhost:9081/openapi/ui/

Reviewer: Can you double-check that this line is correct please? When I download the starter app from openliberty.io, it looks like the default http port is `9080`, not `9081`.

Author: Updated 2e4154864bdcc04bb7f7051e8c4f7dcefc1d0d67

For a fully functional example, check out this link:https://github.com/sweettypdevassy/openAPI-blog[GitHub Repository]. This blog will use this example project for further explanations. 

This is how it appears on the Swagger UI
[.img_border_light]
image::img/blog/OPENAPI-1.png[Swagger UI ,width=70%,align="center"]

> == OpenAPI Annotations 
> 
> MicroProfile OpenAPI provides the following annotations:
> 
> .OpenAPI Annotations
> [options="header"]
> |===
> | Annotation  | Purpose

Reviewer: The annotations come from MicroProfile, they're not part of OpenAPI.

```suggestion
== MicroProfile OpenAPI Annotations 

MicroProfile OpenAPI provides the following annotations:

.MicroProfile OpenAPI Annotations
[options="header"]
|===
| Annotation  | Purpose
```

----
[.img_border_light]
image::img/blog/OPENAPI-3.png[API Metadata Definitions ,width=70%,align="center"] 

*2 . Security Definitions*  

These annotations should used with a REST Application class.

* *`@SecurityScheme`* - Defines authentication for the API.  

* *`@SecuritySchemes`*- Specifies multiple security schemes if needed.  

[source,  java]
----
@SecuritySchemes({ 
    @SecurityScheme( 
        securitySchemeName = "jwtAuth", 
        description = "JWT Authentication", 
        type = SecuritySchemeType.HTTP, 
        scheme = "bearer", 
        bearerFormat = "JWT" 
    ) 
}) 
public class Endpoints extends Application {
----
[.img_border_light]
image::img/blog/OPENAPI-4.png[Security Definitions ,width=70%,align="center"] 
*3 . API Operations*

These annotations should be used with REST resource methods.

* *`@Operation`* - Describes an API operation.  

* *`@APIResponse`* - Defines possible responses for an operation.  

* *`@ApiResponses`* - Groups multiple @APIResponse annotations.  

* *`@Content`* - Specifies the media type and structure of request/response bodies.  

* *`@Schema`* - Defines the properties of a model used in API requests and responses.

> [source,  java]
> ----
> public class Endpoints extends Application {
> @GET
>     @Path("/user/{id}")
>     @Produces(MediaType.APPLICATION_JSON)
>     @Tag(name = "Blog")
>     @Operation(summary = "Get User by ID", description = "Returns the user details for a given ID")
>     @APIResponses({
>     @APIResponse(responseCode = "200", description = "Successful Response", content = @Content(
>         mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = User.class)
>     )),
>     @APIResponse(responseCode = "404", description = "User not found")})
> ----

Reviewer: These annotations are attached to the method, not the class. We need to include the method signature (or at least the start of it)

Also correct the indentation.

```suggestion
[source,  java]
----
@GET
@Path("/user/{id}")
@Produces(MediaType.APPLICATION_JSON)
@Tag(name = "Blog")
@Operation(summary = "Get User by ID", description = "Returns the user details for a given ID")
@APIResponses({
@APIResponse(responseCode = "200", description = "Successful Response", content = @Content(
    mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = User.class)
)),
@APIResponse(responseCode = "404", description = "User not found")})
public User getUser(
----
```

[.img_border_light]
image::img/blog/OPENAPI-5.png[API Operations ,width=70%,align="center"] 
[.img_border_light]
image::img/blog/OPENAPI-6.png[API Operations Response ,width=70%,align="center"] 

*4 . Request Parameter Definitions*  

This annotation should be used with REST resource method parameter.

* *`@Parameter`* - Describes request parameters. 

> [source,  java]
> ----
> public User getUser(
>         @Parameter(description = "User ID", required = true) 
>         @PathParam("id") String id) {
>         
>         return new User(id, "John Doe");
>     }
> ----

Reviewer: Indentation

```suggestion
[source,  java]
----
public User getUser(
        @Parameter(description = "User ID", required = true) 
        @PathParam("id") String id) {
        
    return new User(id, "John Doe");
}
----
```

    public String name; 
----

[.img_border_light]
image::img/blog/OPENAPI-8.png[Schema Definitions ,width=70%,align="center"] 

While basic API documentation is automatically generated from REST methods in Open Liberty, OpenAPI annotations help enhance and refine it. This is similar to how Javadoc comments work in Java—without them, Javadoc can still list methods, parameters, and return types, but it lacks detailed explanations of behavior and usage. Likewise, OpenAPI annotations provide essential context to REST endpoints, describing their purpose, expected inputs, responses, and error handling. Additionally, some API details cannot be inferred directly from Java code, such as custom response headers, security requirements, or detailed descriptions of parameters. In such cases, annotations become essential for ensuring comprehensive, well-structured API documentation that is both developer-friendly and machine-readable.

The MicroProfile OpenAPI feature is responsible for generating this documentation.This eliminates the need for external configuration files and ensures the documentation evolves alongside the codebase.

> Ultimately, OpenAPI annotations play a vital role in creating useful, up-to-date documentation. Because they reside within the source code, they make it easier for teams to maintain alignment between implementation and documentation—benefiting both developers and end users who need to understand how to interact with the API effectively.

Reviewer: The annotations come from MicroProfile, they're not part of OpenAPI.

```suggestion
While basic API documentation is automatically generated from REST methods in Open Liberty, MicroProfile OpenAPI annotations help enhance and refine it. This is similar to how Javadoc comments work in Java—without them, Javadoc can still list methods, parameters, and return types, but it lacks detailed explanations of behavior and usage. Likewise, MicroProfile OpenAPI annotations provide essential context to REST endpoints, describing their purpose, expected inputs, responses, and error handling. Additionally, some API details cannot be inferred directly from Java code, such as custom response headers, security requirements, or detailed descriptions of parameters. In such cases, annotations become essential for ensuring comprehensive, well-structured API documentation that is both developer-friendly and machine-readable.

The MicroProfile OpenAPI feature is responsible for generating this documentation. This eliminates the need for external configuration files and ensures the documentation evolves alongside the codebase.

Ultimately, MicroProfile OpenAPI annotations play a vital role in creating useful, up-to-date documentation. Because they reside within the source code, they make it easier for teams to maintain alignment between implementation and documentation—benefiting both developers and end users who need to understand how to interact with the API effectively.
```

> The MicroProfile OpenAPI feature is responsible for generating the documentation, providing a structured way to describe the API. To enhance this documentation, developers can use MicroProfile OpenAPI annotations, which allow them to add more details and improve clarity. Having comprehensive and well-maintained documentation is highly beneficial for end users, as it helps them understand how to interact with the API effectively. These annotations are embedded directly within the code, alongside the implementation, ensuring that the documentation remains closely tied to the source. This approach makes it easier for developers to update and maintain the documentation as they refine and expand the codebase over time.

Reviewer: I think the first paragraph here is really good but the three paragraphs repeat each other a lot.

Author: Updated 5c3df8b62a74be1e48b06a4513e8b8a53bd920bb

Reviewer: This looks better, thanks.

== Influence  
As *API architectures evolve,* OpenAPI, the industry-standard specification for *documenting REST APIs,* is influencing how *GraphQL* and *gRPC* APIs handle documentation and interoperability. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability. 

* OpenAPI’s structured *JSON/YAML format* has influenced *GraphQL’s SDL (Schema Definition Language),* where types, queries, and mutations are explicitly defined. 

> * Inspired by OpenAPI's *Swagger UI*, gRPC now has *gRPC UI*, which allows developers to explore and test API methods interactively, much like Swagger does for REST APIs.   

Reviewer: This sentence makes it sound like Swagger UI is part of the OpenAPI project, and gRPC UI is part of gRPC, whereas I think both are separate projects?

Admittedly, OpenAPI grew out of Swagger, and both Swagger and Swagger UI were created by Smartbear.

Author: Updated ef556c328e3eb6dc201be68c43f963c4e6307d8e

Reviewer: I think that's clearer, thanks.

> * Inspired by OpenAPI's *Swagger UI*, gRPC now has *gRPC UI*, which allows developers to explore and test API methods interactively, much like Swagger does for REST APIs.   

Reviewer: I can't review this section as I'm not familiar with GraphQL or gRPC.
